lib/ansible/parsing/vault/__init__.py

```diff
@@ -104,116 +104,132 @@ def check_prereqs():
     if not HAS_AES or not HAS_COUNTER or not HAS_ANY_PBKDF2HMAC or not HAS_HASH:
         raise AnsibleError(CRYPTO_UPGRADE)
 
-class VaultLib(object):
+class VaultLib:
 
     def __init__(self, password):
-        self.password = password
+        self.b_password = to_bytes(password, errors='strict', encoding='utf-8')
```

Comment:
Probably not, but passwords are a little special so a case could be made:   In general we want our strategy to be unicode type internally and bytes type when we actually use it.  That way if we manipulate the string in some way (measure its length, concatenate it, strip whitespace characters, etc) there's no problems with bytes doing the wrong thing or getting mixed with unicode.  By keeping everything unicode type internally there's no confusion of what type you're dealing with and it keeps unnecessary conversions to a minimum (to_bytes and to_unicode are pretty good about unnecessary conversions but we still have the function call overhead and a few conditionals everytime we call it.)

For passwords, we do very few of these things.  So it may make sense to carry the byte type all the way through from input to output.  However, that requires tracing a lot of code, finding all the places we input vault passwords and making sure that they maintain the proper type all the way through.

Since the vault code that uses the password as a byte string is self-contained inside or behind VaultLib, I think it makes sense to do this transformation here and then check the rest of the code at our leisure later.